internal/oapispec/ffi2swagger.go

```diff
@@ -35,35 +38,101 @@ func NewFFISwaggerGen() FFISwaggerGen {
 	return &ffiSwaggerGen{}
 }
 
-func (og *ffiSwaggerGen) Generate(ctx context.Context, baseURL string, ffi *fftypes.FFI) (swagger *openapi3.T, err error) {
-
-	// routes := []*Route{}
-
-	// for _, method := range ffi.Methods {
-	// 	if routes, err = og.addMethod(ctx, routes, method); err != nil {
-	// 		return nil, err
-	// 	}
-	// }
-
-	// for _, event := range ffi.Events {
-	// 	if routes, err = og.addEvent(ctx, routes, event); err != nil {
-	// 		return nil, err
-	// 	}
-	// }
-
-	return nil, nil
+func (og *ffiSwaggerGen) Generate(ctx context.Context, baseURL string, api *fftypes.ContractAPI, ffi *fftypes.FFI) (swagger *openapi3.T, err error) {
+	hasLocation := api.Location.String() != "null"
+
+	routes := []*Route{}
+	for _, method := range ffi.Methods {
+		routes = og.addMethod(routes, method, hasLocation)
+	}
+	for _, event := range ffi.Events {
+		routes = og.addEvent(routes, event, hasLocation)
+	}
+
+	return SwaggerGen(ctx, routes, &SwaggerGenConfig{
+		Title:       ffi.Name,
+		Version:     ffi.Version,
+		Description: ffi.Description,
+		BaseURL:     baseURL,
+	}), nil
 }
 
-// func (og *ffiSwaggerGen) addMethod(ctx context.Context, routes []*Route, method *fftypes.FFIMethod) ([]*Route, error) {
+func (og *ffiSwaggerGen) addMethod(routes []*Route, method *fftypes.FFIMethod, hasLocation bool) []*Route {
+	return append(routes, &Route{
+		Name:             fmt.Sprintf("invoke_%s", method.Pathname),
+		Path:             fmt.Sprintf("invoke/%s", method.Pathname),
+		Method:           http.MethodPost,
+		JSONInputValue:   func() interface{} { return &fftypes.InvokeContractRequest{} },
+		JSONInputSchema:  func(ctx context.Context) string { return invokeRequestJSON(&method.Params, hasLocation).String() },
+		JSONOutputValue:  func() interface{} { return &fftypes.JSONObject{} },
+		JSONOutputSchema: func(ctx context.Context) string { return ffiParamsJSON(&method.Returns).String() },
+		JSONOutputCodes:  []int{http.StatusOK},
+	})
+}
 
-// 	routes = append(routes, &Route{
-// 		Name: fmt.Sprintf("method_%s_invoke", method.Name),
-// 	})
+func (og *ffiSwaggerGen) addEvent(routes []*Route, event *fftypes.FFIEvent, hasLocation bool) []*Route {
+	// If the API has a location specified, there are no fields left to specify in the request body.
+	// Instead of masking them all (which causes Swagger UI some issues), explicitly set the schema to an empty object.
+	var schema func(ctx context.Context) string
+	if hasLocation {
+		schema = func(ctx context.Context) string { return `{"type": "object"}` }
+	}
+	return append(routes, &Route{
+		Name:            fmt.Sprintf("subscribe_%s", event.Pathname),
+		Path:            fmt.Sprintf("subscribe/%s", event.Pathname),
```

Comment:
We had previously said this should be a `/subscriptions` collection, with CRUD operations (which are filtered versions of the CRUD operations on `/subscriptions ` the top-level).
I'm probably just behind on how things evolved, but wanted to highlight the gap in my understanding.

Reply:
Yea, that was the plan, but the number of URIs started to get out of hand and the filtering became quite confusing. One major problem was since an API can be replaced with a PUT, it was unclear what to do with "child" subscriptions (delete them? replace them?)

Where we've currently landed is that the CRUD operations for subscriptions are only at the top level, but there are RPC-style "/subscribe" helpers at both the FFI and API level, as an alternate way of creating subscriptions with some of the data pre-filled.